### PR TITLE
[AArch64][GISel] Support SVE with 128-bit min-size for G_LOAD and G_STORE

### PR DESCRIPTION
**Pull request by Him188**
This patch adds basic support for scalable vector types in load & store instructions for AArch64 with GISel.

Only scalable vector types with a 128-bit base size are supported, e.g. `<vscale x 4 x i32>`, `<vscale x 16 x i8>`.

This patch adapted some ideas from a similar abandoned patch [https://github.com/llvm/llvm-project/pull/72976](https://github.com/llvm/llvm-project/pull/72976).

### CODE REVIEW

**Comment by Him188** [comment]
Requesting review from @sjoerdmeijer, @davemgreen, @tschuett, @arsenm

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 1]
```diff
@@ -2906,9 +2927,23 @@ bool AArch64InstructionSelector::select(MachineInstr &I) {
...
+#ifndef NDEBUG
+    if (ValTy.isScalableVector()) {
+        assert(STI.hasSVE() 
+             && "Load/Store register operand is scalable vector "
+                "while SVE is not supported by the target");
+        // assert(RB.getID() == AArch64::SVRRegBankID 
+        //        && "Load/Store register operand is scalable vector "
+        //           "while its register bank is not SVR");
+    }
+#endif
```
Comment (dc03-work) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:2939]: You can remove the `#ifndef` and the commented assert (asserts are automatically disabled when  `LLVM_ENABLE_ASSERTIONS` is set to `OFF`)

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -26375,12 +26375,20 @@ bool AArch64TargetLowering::shouldLocalize(
...
+  return Op == Instruction::Load || Op == Instruction::Store;
+}
+
 bool AArch64TargetLowering::fallBackToDAGISel(const Instruction &Inst) const {
-  if (Inst.getType()->isScalableTy())
-    return true;
+  const auto ScalableTySupported = isScalableTySupported(Inst.getOpcode());
+
+  // Fallback for scalable vectors
+  if (Inst.getType()->isScalableTy() && !ScalableTySupported) {
```
Comment (dc03-work) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:26386]: We also need to distinguish between GlobalISel and FastISel here. FastISel also calls this function, and it doesn't support scalable vectors. The patch you linked uses a command line flag to do this, but there is probably a better way than that.

File: llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll [hunk 3]
```diff
@@ -0,0 +1,50 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 4
+; RUN: llc -mtriple=aarch64-linux-gnu -mattr=+sve -global-isel < %s | FileCheck %s
+
+define void @scalable_v16i8(ptr noalias nocapture noundef %l0, ptr noalias nocapture noundef %l1) {
```
Comment (dc03-work) [inline llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:4]: The attributes on the function parameters are not required

Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll]: Removed the attributes

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 4]
```diff
@@ -61,6 +61,79 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
...
   const LLT v2p0 = LLT::fixed_vector(2, p0);
 
+  // Scalable vector sizes range from 128 to 2048
+  // Note that subtargets may not support the full range.
+  // See [ScalableVecTypes] below.
+  const LLT nxv16s8 = LLT::scalable_vector(16, s8);
+  const LLT nxv32s8 = LLT::scalable_vector(32, s8);
+  const LLT nxv64s8 = LLT::scalable_vector(64, s8);
+  const LLT nxv128s8 = LLT::scalable_vector(128, s8);
+  const LLT nxv256s8 = LLT::scalable_vector(256, s8);
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:71]: This is a scalable vector, which size is an unknown multiple of 256 * 8. Legalizing loads of scalable vectors is unintuitive.

**Comment by Him188** [comment]
> You cannot use `legalForTypesWithMemDesc` for legalizing scalable vector loads?

I intended to reduce the manual permutation of the scalable vector types. We also need to check whether ST supports such vector sizes. 

Since I have already generated `ScalableVecTypes` based on ST capabilities, I decided to reuse that information. Will it be better if I generate rules for `legalForTypesWithMemDesc` instead?

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 4 above)
Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:71]: As far as I know, there is no target that supports `nxv256s8`. The original patch only listed:
```
 const LLT nxv8s16 = LLT::scalable_vector(8, 16);
 const LLT nxv4s32 = LLT::scalable_vector(4, 32);
 const LLT nxv2s64 = LLT::scalable_vector(2, 64);
```

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 5]
```diff
@@ -61,6 +61,79 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
...
+    nxv256s8, nxv128s16, nxv64s32, nxv32s64, nxv32p0,
+  };
+
+  /// Scalable vector types supported by the sub target.
+  /// Empty if SVE is not supported.
+  SmallVector<LLT> ScalableVecTypes;
+  
+  if (ST.hasSVE()) {
+    // Add scalable vector types that are supported by the subtarget
+    const auto MinSize = ST.getMinSVEVectorSizeInBits();
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:119]: I thought the point of SVE is that you can compile on one machine and run on another with a different vector widths.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:26386]: Hi - Please keep the command line option from #72976 at the moment. It will help having a debug option we can use for testing, but not be limited to implement every single SVE optimization before global-isel becomes production ready.

This doesn't need to be dependant on the operations we currently "support", the normal fallback mechanism for unsupported operations should handle that.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 6]
```diff
@@ -901,6 +901,27 @@ static unsigned selectLoadStoreUIOp(unsigned GenericOpc, unsigned RegBankID,
   return GenericOpc;
 }
 
+/// Select the AArch64 opcode for the G_LOAD or G_STORE operation for scalable 
+/// vectors.
+/// \p ElementSize size of the element of the scalable vector
+static unsigned selectLoadStoreSVEOp(const unsigned GenericOpc,
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:907]: Is this manually selected in the DAG path? Can you do this in tablegen?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 7]
```diff
@@ -26375,12 +26375,20 @@ bool AArch64TargetLowering::shouldLocalize(
...
+      return true;
+  }
 
   for (unsigned i = 0; i < Inst.getNumOperands(); ++i)
-    if (Inst.getOperand(i)->getType()->isScalableTy())
+    if (Inst.getOperand(i)->getType()->isScalableTy() && !ScalableTySupported)
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:26391]: The condition is getting bigger. Please take them out, assign to booleans and check.

File: llvm/lib/Target/AArch64/AArch64RegisterBanks.td [hunk 8]
```diff
@@ -13,7 +13,7 @@
 def GPRRegBank : RegisterBank<"GPR", [XSeqPairsClass]>;
 
 /// Floating Point/Vector Registers: B, H, S, D, Q.
-def FPRRegBank : RegisterBank<"FPR", [QQQQ]>;
+def FPRRegBank : RegisterBank<"FPR", [QQQQ, ZPR]>;
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/AArch64RegisterBanks.td:16]: The comment needs update to include ZPR too.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 9]
```diff
@@ -1413,7 +1413,7 @@ bool IRTranslator::translateLoad(const User &U, MachineIRBuilder &MIRBuilder) {
 
 bool IRTranslator::translateStore(const User &U, MachineIRBuilder &MIRBuilder) {
   const StoreInst &SI = cast<StoreInst>(U);
-  if (DL->getTypeStoreSize(SI.getValueOperand()->getType()) == 0)
+  if (DL->getTypeStoreSize(SI.getValueOperand()->getType()).isZero())
```
Comment (madhur13490) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1416]: This seems unrelated change. Isn't it?

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 6 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:907]: https://github.com/llvm/llvm-project/blob/e2ec9514876744627e99fecb7179accdc6969e4e/llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp#L865-L902

Currently we are already doing `selectLoadStoreUIOp` manually. I'm following the same idea.
Making them select through TableGen would require more work and can be a separate patch.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 9 above)
Comment (Him188) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:1416]: It's needed, otherwise we implicitly cast a ScalableVectorTy into ScalarTy and will fail.
```c++
TypeSize::operator TypeSize::ScalarTy() const {
  if (isScalable()) {
    reportInvalidSizeRequest(
        "Cannot implicitly convert a scalable size to a fixed-width size in "
        "`TypeSize::operator ScalarTy()`");
    return getKnownMinValue();
  }
  return getFixedValue();
}
```

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 4 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:71]: To make it simpler for now, I would remove all these except for base sizes of 128-bit:
- `nxv8s16`, `nxv4s32`, and `nxv2s64` as listed in the original patch
- `nxv16s8` since SelectionDAG supports it
- `nxv2p0` since it has the same effective size as `nxv2s64`.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 5 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:119]: Agreed, checking `ST.hasSVE()` is enough, at least for sizes of multiple of 128-bit

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:119]: There is a `const bool HasRCPC3 = ST.hasRCPC3();` in the file. You could add a similar:
```
const bool HasSVE = ST.hasSVE();
```

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 10]
```diff
@@ -740,11 +741,14 @@ AArch64RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
...
       continue;
-    OpSize[Idx] = Ty.getSizeInBits();
+    OpSize[Idx] = Ty.getSizeInBits().getKnownMinValue();
 
-    // As a top-level guess, vectors go in FPRs, scalars and pointers in GPRs.
+    // As a top-level guess, scalable vectors go in SVRs, non-scalable
+    // vectors go in FPRs, scalars and pointers in GPRs.
     // For floating-point instructions, scalars go in FPRs.
-    if (Ty.isVector() || isPreISelGenericFloatingPointOpcode(Opc) ||
+    if (Ty.isScalableVector())
```
Comment (aemerson) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:749]: Isn't `isScalableVector()` already covered by `isVector()`?

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 11]
```diff
@@ -740,11 +741,14 @@ AArch64RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
...
+    OpSize[Idx] = Ty.getSizeInBits().getKnownMinValue();
 
-    // As a top-level guess, vectors go in FPRs, scalars and pointers in GPRs.
+    // As a top-level guess, scalable vectors go in SVRs, non-scalable
+    // vectors go in FPRs, scalars and pointers in GPRs.
     // For floating-point instructions, scalars go in FPRs.
-    if (Ty.isVector() || isPreISelGenericFloatingPointOpcode(Opc) ||
+    if (Ty.isScalableVector()) 
+      OpRegBankIdx[Idx] = PMI_FirstFPR;
+    else if (Ty.isVector() || isPreISelGenericFloatingPointOpcode(Opc) ||
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:751]: We have to distinguish between scalable and fixed sized vectors explicitly.
```
 Ty.getSizeInBits() 
```
asserts for scalable vectors.

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:751]: or put all vectors into the first if. The second is for > 64 bits.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 12]
```diff
@@ -329,6 +335,20 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
     return ValTy.isPointerVector() && ValTy.getAddressSpace() == 0;
   };
 
+  if (ST.hasSVE()) {
+    for (const auto OpCode : {G_LOAD, G_STORE}) {
+      getActionDefinitionsBuilder(OpCode)
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:340]: IIRC, you cannot do `getActionDefinitionsBuilder` for the same opcode twice, SVE and non-SVE.

Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:340]: Oh..

Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:340]: Look for G_ABS.

Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:340]:
```C++
LegalizeRuleSet &LegalizerInfo::getActionDefinitionsBuilder(unsigned Opcode) {
  unsigned OpcodeIdx = getActionDefinitionsIdx(Opcode);
  auto &Result = RulesForOpcode[OpcodeIdx];
  assert(!Result.isAliasedByAnother() && "Modifying this opcode will modify aliases");
  return Result;
}
```
`getActionDefinitionsBuilder` just returns a reference so I think doing it twice is not a problem.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 7 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:26391]: Added an option, so this check is no longer needed.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 12 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp]: Changed accordingly

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 13]
```diff
@@ -145,6 +145,15 @@ static cl::opt<bool> EnableExtToTBL("aarch64-enable-ext-to-tbl", cl::Hidden,
 static cl::opt<unsigned> MaxXors("aarch64-max-xors", cl::init(16), cl::Hidden,
                                  cl::desc("Maximum of xors"));
 
+// By turning this on, we will not fallback to DAG ISel when encountering
+// scalable vector types for all instruction, even if SVE is not yet supported
+// with some instructions.
+// See [AArch64TargetLowering::fallbackToDAGISel] for implementation details.
+static cl::opt<bool> EnableSVEGISel(
+    "aarch64-enable-sve-gisel", cl::Hidden,
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:153]: Better would be -aarch64-enable-gisel-sve.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 14]
```diff
@@ -329,7 +335,23 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
     return ValTy.isPointerVector() && ValTy.getAddressSpace() == 0;
   };
 
-  getActionDefinitionsBuilder(G_LOAD)
+  auto &LoadActions = getActionDefinitionsBuilder(G_LOAD);
+  auto &StoreActions = getActionDefinitionsBuilder(G_STORE);
+
+  if (ST.hasSVE()) {
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:341]: For this PR, I vote for the G_ABS pattern and remove the verbose loop. We should open a ticket to discuss making feature-based legalization more economic. The X86 guys have similar issues.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 15]
```diff
@@ -2906,9 +2927,18 @@ bool AArch64InstructionSelector::select(MachineInstr &I) {
...
+             && "Load/Store register operand is scalable vector "
+                "while SVE is not supported by the target");
+    }
+    
     // The code below doesn't support truncating stores, so we need to split it
     // again.
-    if (isa<GStore>(LdSt) && ValTy.getSizeInBits() > MemSizeInBits) {
+    // Truncate only if type is not scalable vector
+    const bool NeedTrunc = !ValTy.isScalableVector() 
+                      && ValTy.getSizeInBits().getFixedValue() > MemSizeInBits;
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:2940]: Should I add truncate / zext / sext in this patch?

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:341]: If we remove the loop then we have to duplicate the patterns for both G_LOAD and G_STORE?

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:2940]: That can be separate - one step at a time, and it would be good to get the basics in first :)

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:341]: I think that might be OK - they might end up being different eventually anyway, after all the truncstores/extloads and type legalization gets added.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 16]
```diff
@@ -329,6 +335,20 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
     return ValTy.isPointerVector() && ValTy.getAddressSpace() == 0;
   };
 
+  if (ST.hasSVE()) {
+    for (const auto OpCode : {G_LOAD, G_STORE}) {
+      getActionDefinitionsBuilder(OpCode)
+      .legalForTypesWithMemDesc({
+        // 128 bit base sizes
+        {nxv16s8, p0, nxv16s8, 128},
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:345]: Is it appropriate to allow a minimum of 8? 128 seems to be more efficient but 8 should also work.

File: llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+; CHECK-NEXT:    ptrue p0.s
+; CHECK-NEXT:    ld1w { z0.s }, p0/z, [x0]
+; CHECK-NEXT:    st1w { z0.s }, p0, [x1]
+; CHECK-NEXT:    ret
+  %l3 = load <vscale x 4 x i32>, ptr %l0, align 16
+  store <vscale x 4 x i32> %l3, ptr %l1, align 16
+  ret void
+}
+
+define void @scalable_v2i64(ptr %l0, ptr %l1) {
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: It turns out ptrs do not work after adding the test. 
It failed to match an existing tablegen pattern during instruction selection. SDAG seems to handle ptrs as s64 but I'm not sure how GISEL handles ptrs. Does it also convert ptrs to s64 at some stage or need special instruction selection pattern for that?

BTW, in RISCV they don't define `nxvNp0` types.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 6 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:907]: I did some investigation and found that the patch is actually using the existing tablegen-ed patterns in  `selectImpl`. The added instruction selector code is just skipped.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: > Please format the entire patch.

All formatting done

File: llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h [hunk 18]
```diff
@@ -652,17 +652,17 @@ bool GIMatchTableExecutor::executeMatchTable(
       MachineMemOperand *MMO =
           *(State.MIs[InsnID]->memoperands_begin() + MMOIdx);
 
-      unsigned Size = MRI.getType(MO.getReg()).getSizeInBits();
+      const auto Size = MRI.getType(MO.getReg()).getSizeInBits();
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/include/llvm/CodeGen/GlobalISel/GIMatchTableExecutorImpl.h]: done

File: llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll (hunk 17 above)
[the remark this answers is not in the document]
Comment (dc03-work) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: GlobalISel uses `p<n>` (only `p0` is used right now?) throughout the pipeline, afaik its not converted to integers anywhere.

[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: AArch64 only uses p0. It is as pointer in address space 0. If we cannot select nxv2p0, then we should not legalize it.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp:344]: Yeah 8 should be fine. SDAG seems to accept the smaller alignments.

File: llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll (hunk 17 above)
[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: There are two options for ptrs that are stored
- Scalars are copied it to integer in AArch64InstructionSelector::preISelLower
- Vectors are bitcast to integer vectors in AArch64LegalizerInfo::legalizeLoadStore, by marking them as custom.
We only really need one method though. The vector stores should be using bitcastIf not customIf and the scalars could hopefully use the same method. This might mean that G_BITCAST needed to be supported for scalable vectors though, if it was using the same method.

[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: I vote for postponing support for nxv2p0. The patch is large enough. Bitcasting seems to be preferable.

[the remark this answers is not in the document]
Comment (davemgreen) [reply llvm/test/CodeGen/AArch64/GlobalISel/sve-load-store.ll:40]: Yeah that sounds OK to me. Perhaps add a TODO about scalable vector ptr stores, so we can look into it later.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 19]
```diff
@@ -2853,8 +2874,8 @@ bool AArch64InstructionSelector::select(MachineInstr &I) {
       return false;
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: I reverted the entire file AArch64InstructionSelector.cpp because we can reuse tablegen patterns. So I think we don't need to change here.

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp [hunk 20]
```diff
@@ -440,8 +463,7 @@ AArch64LegalizerInfo::AArch64LegalizerInfo(const AArch64Subtarget &ST)
           {p0, v4s32, v4s32, 8},
           {p0, v2s64, v2s64, 8},
           {p0, v2p0, v2p0, 8},
-          {p0, s128, s128, 8},
-      })
+          {p0, s128, s128, 8}})
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp]: Done, formatted the entire patch.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 21]
```diff
@@ -2906,9 +2927,18 @@ bool AArch64InstructionSelector::select(MachineInstr &I) {
     const LLT ValTy = MRI.getType(ValReg);
     const RegisterBank &RB = *RBI.getRegBank(ValReg, MRI, TRI);
 
+    if (ValTy.isScalableVector()) {
+        assert(STI.hasSVE() 
+             && "Load/Store register operand is scalable vector "
+                "while SVE is not supported by the target");
+    }
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp]: Reverted this change because we can reuse tablegen pattern and will not reach here

File: llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp]: Ok, changed to 8

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp [hunk 22]
```diff
@@ -309,7 +309,7 @@ bool matchSplitStoreZero128(MachineInstr &MI, MachineRegisterInfo &MRI) {
   if (!Store.isSimple())
     return false;
   LLT ValTy = MRI.getType(Store.getValueReg());
-  if (!ValTy.isVector() || ValTy.getSizeInBits() != 128)
+  if (!ValTy.isVector() || ValTy.getSizeInBits().getKnownMinValue() != 128)
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp]: Done, reverted this line and added this above instead.
```cpp
if (ValTy.isScalableVector())
    return false;
```

File: llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp [hunk 23]
```diff
@@ -657,8 +657,8 @@ bool AArch64PostLegalizerCombiner::optimizeConsecutiveMemOpAddressing(
         Register PtrBaseReg;
         APInt Offset;
         LLT StoredValTy = MRI.getType(St->getValueReg());
-        unsigned ValSize = StoredValTy.getSizeInBits();
-        if (ValSize < 32 || St->getMMO().getSizeInBits() != ValSize)
+        const auto ValSize = StoredValTy.getSizeInBits();
```
[the remark this answers is not in the document]
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64PostLegalizerCombiner.cpp]: Done, replaced with a bail-out check for operand 0 being scalable vector.

**Review by aemerson — APPROVED** [review]
Thanks for working on this as a first step. LGTM now.

**Review by davemgreen — APPROVED** [review]
Lets do it. LGTM

**Comment by Him188** [comment]
On the failed CI: 
By searching `FAIL:` there is only one match, `OpenMP/target_ast_print.cpp`. 
I checked that for the commit this PR is based on, this test is also failing. Should we ignore this or rebase the PR?

**Comment by aemerson** [comment]
> On the failed CI: By searching `FAIL:` there is only one match, `OpenMP/target_ast_print.cpp`. I checked that for the commit this PR is based on, this test is also failing. Should we ignore this or rebase the PR?

Yes you can just ignore that.
